web/src/features/public-api/server/RateLimitService.ts

```diff
@@ -0,0 +1,216 @@
+    resource: z.infer<typeof RateLimitResource>,
+  ) {
+    const planBasedConfig = getRateLimitConfig(scope.plan, resource);
+
+    const customConfig = scope.rateLimits?.find(
+      (config) => config.resource === resource,
+    );
+
+    const effectiveConfig = customConfig || planBasedConfig;
+
+    // returning early if no rate limit is set
+    if (
+      !effectiveConfig ||
+      !effectiveConfig.points ||
+      !effectiveConfig.durationInSec
+    ) {
+      return;
+    }
+
+    const rateLimiter = new RateLimiterRedis({
+      // Basic options
+      points: effectiveConfig.points, // Number of points
+      duration: effectiveConfig.durationInSec, // Per second(s)
+
+      keyPrefix: this.rateLimitPrefix(resource), // must be unique for limiters with different purpose
+      storeClient: this.redis,
+      rejectIfRedisNotReady: true,
+    });
+
+    let res: RateLimitResult | undefined = undefined;
+    try {
+      // orgId used as key for different resources
+      const libRes = await rateLimiter.consume(scope.orgId);
+      res = {
+        resource,
+        scope,
+        points: effectiveConfig.points,
+        remainingPoints: libRes.remainingPoints,
+        msBeforeNext: libRes.msBeforeNext,
+        consumedPoints: libRes.consumedPoints,
+        isFirstInDuration: libRes.isFirstInDuration,
+      };
+    } catch (err) {
+      if (err instanceof RateLimiterRes) {
+        // No points available or key is blocked
+        res = {
+          resource,
+          scope,
+          points: effectiveConfig.points,
+          remainingPoints: err.remainingPoints,
+          msBeforeNext: err.msBeforeNext,
+          consumedPoints: err.consumedPoints,
+          isFirstInDuration: err.isFirstInDuration,
+        };
+      } else {
+        // Some other error occurred, rethrow it
+        console.log("Internal Rate limit error", err);
+        throw err;
+      }
+    }
+
+    if (res.remainingPoints < 1) {
+      recordIncrement("rate_limit_exceeded", 1, {
+        orgId: scope.orgId,
+        plan: scope.plan,
+        resource: resource,
+      });
+    }
+
+    return res;
+  }
+
+  rateLimitPrefix(resource: string) {
+    return `rate-limit:${resource}`;
+  }
+}
+
+export class RateLimitHelper {
+  res: RateLimitResult | undefined;
+
+  constructor(res: RateLimitResult | undefined) {
+    this.res = res;
+  }
+
+  isRateLimited() {
+    return this.res ? this.res.remainingPoints < 1 : false;
+  }
+
+  sendRestResponseIfLimited(nextResponse: NextApiResponse) {
+    if (!this.res || !this.isRateLimited()) {
+      return;
+    }
+    return sendRateLimitResponse(nextResponse, this.res);
+  }
+}
+
+export const sendRateLimitResponse = (
+  res: NextApiResponse,
+  rateLimitRes: RateLimitResult,
+) => {
+  const httpHeader = createHttpHeaderFromRateLimit(rateLimitRes);
+
+  for (const [header, value] of Object.entries(httpHeader)) {
+    res.setHeader(header, value);
+  }
+
+  res.status(429).end();
```

Comment:
shall we add a human readable response in addition to status and headers? e.g. simple "429 - rate limit exceeded"